solutions/gold/cf-1187E.mdx

```diff
@@ -0,0 +1,172 @@
+the rest of the tree as a part of the leftover connected component. This hints
+for a DP on Trees solution, we choose one leaf node, work our way up and work
+our way down for each of the siblings. For example, for the following tree:
+
+![Tree Visualization](cf-1187E/tree.png)
+
+If we select the 4th node as our starting point, then the process would look
+like the following:
+
+1. Work our way up to 3
+2. Work our way down to 5
+3. Work our way up to 1
+4. Work our way down to 2
+
+We need to maintain the cost of two different scenarios: working up and working
+down, meaning that we have to maintain 2 DP arrays. Let's define the DP
+transitions for each of the arrays. We will use $dp_1$ for the cost of working
+down and $dp_2$ for the cost of work up and $sz$ for the size of a subtree.
+
+$$
+sz[u] = 1 + \sum_{c\in child(u)} sz[c] \newline
+dp_1[u] = sz[u] + \sum_{c\in child(u)} dp_1[c]
+$$
+
+We can calculate this DP array with a single DFS search. Now, for the second DP
+array:
+
+$$
+dp_2[u] = \max_{c\in child(u)} (sz[u] - sz[c] + dp_2[c] + \sum_{d\in child(u), c\neq d}dp_1[d])
+$$
+
+There is still one problem though. The cost of $sz[u] - sz[c]$ is applied the
+number of times the initially chosen node is deep compared to $u$. However, we
+can't anticipate this number beforehand, and we might have some ancestor node
+later on which would have prefered a higher depth over a higher cost. This means
+that we need to incorporate the $sz[u] - sz[c]$ for every ancestor node when
+initializing the $dp_2$ value for leaf nodes. That way we, it will be
+incorporated into the cost beforehand and we don't need to worry about choosing
+the wrong initial node with an inoptimal ending cost.
+
+We can calculate these initial values with another DFS search before the one
+calculating the values for $dp_2$. That way we can remove the $sz[u] - sz[c]$
+from the recurrence formula for $dp_2$. To calculate these initial values, we
+notice that every time we go down by one depth, we add size of all the siblings
+and parent combined two times, one for the parent, and once for the current
+node. We also add the cost of all the ancestor's siblings and parent combined
+once more. Let's say $p_2[u]$ is the initial cost for a leaf node $u$, and
+$p_1[u]$ is the cost of the ancestor's siblings and parent combined till $u$.
+Then we can calculate $p_1[u]$ and $p_2[u]$ using the following recurrence:
+
+$$
+p_1[u] = p_1[p] + sz[p] - sz[u] \newline
+p_2[u] = p_2[p] + p_1[u] + sz[p] - sz[u]
+$$
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(V+E)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+
+vector<vector<int>> g;
+vector<ll> dp1, dp2;
+vector<ll> p1, p2;
+vector<ll> sz;
+
+// Calculate sz and dp1
+void dfs1(int u, int p) {
+	sz[u] = 1;
+	for (int &v: g[u]) {
+		if (v == p) {
+			continue;
+      dfs1(v, u);
+		}
```

Comment:
sus

Reply:
o lol i was editing in vscode it prob reset the spaces